File: archive/python/mergesort.py

def sort(elements: list[int]) -> list[int]:
    print(f"Sort: {elements}")
    if len(elements) > 1:
        left_half, right_half = split(elements)
        
        left_half_sorted = sort(left_half)
        right_half_sorted = sort(right_half)
 
        merged = merge(left_half_sorted, right_half_sorted)
        print(f"Merged: {merged}")

        return merged
    else:
        return elements
    

def split(elements: list[int]) -> tuple[list[int], list[int]]:
    mid = len(elements) // 2
    left_half = elements[:mid]
    right_half = elements[mid:]
    return left_half, right_half

def merge(left: list[int], right: list[int]):
    if len(left) == 0 or len(right) == 0:
        return left + right

    merged = [] 
    idx = 0
    for l in left:
        i = idx
        for r in right[i:]:
            if r <= l:
                merged.append(r)
            else:
                break
            idx += 1
        merged.append(l)
    
    merged += right[idx:]

    return merged

if __name__ == "__main__":
    sorted = sort([13, 11, 12, 5, 6, 7])
    print(f"Result: {sorted}")

File: archive/c/mergesort.c
#include <stdio.h>
#include <string.h>

int* mergesort(int* sorted, int* elements, size_t count);
int* merge(int* merged, int* left, int* right, size_t leftCount, size_t rightCount);
void printArr(char* text, int* arr, size_t count);


int main() { 
    printf("Starting mergesort.\n");
    int elements[] = {13, 11, 12, 5, 6, 7};
    size_t count = sizeof(elements) / sizeof(*elements);
    int sorted[count];

    mergesort(sorted, elements, count);

    return 0;
}


void printArr(char* text, int* arr, size_t count){
   printf("%s",  text);
   printf("[");

   int indexLast = count - 1;

   if (indexLast >= 0){
        for(int i=0; i<indexLast; i++){
            printf("%i, ", arr[i]);
    }
        printf("%d", arr[indexLast]);
   }
   printf("]\n");
}


int* mergesort(int* sorted, int* elements, size_t count) {
    printArr("Sorting: ", elements, count);
    
    if (count <= 1){
        memcpy(sorted, elements, sizeof(int));
        printArr("Sorted: ", sorted, count);
        return sorted;
    }

    // Split elements into two halves.
    int mid = (int) count / 2;
    int leftCount = mid, rightCount = count-mid;
    int *leftHalf = elements, *rightHalf = elements + mid;

    // Apply mergesort on each half.
    int leftSorted[leftCount], rightSorted[rightCount];
    mergesort(leftSorted, leftHalf, leftCount);
    mergesort(rightSorted, rightHalf, rightCount);

    // Merge the two halves.
    merge(sorted, leftSorted, rightSorted, leftCount, rightCount);
    printArr("Sorted: ", sorted, count);
    return sorted;  
}


int* merge(int* merged, int* left, int* right, size_t leftCount, size_t rightCount) {
    int leftOffset = 0, offset = 0, idx = 0;
    for(int l=0; l<leftCount; l++){
        for(int r=offset; r<rightCount; r++){
            if(left[l] <= right[r]){
                merged[idx] = left[l];
                idx ++;
                leftOffset ++; 
                break;
            }
            merged[idx] = right[r];
            idx ++;
            offset++;
        }
    }
    for(int l=leftOffset; l<leftCount; l++){
        merged[idx] = left[l];
        idx++;
    }
    for(int r= offset; r<rightCount; r++){
        merged[idx] = right[r];
        idx++;
    }

    return merged;
}
